python/thunder/imgprocessing/regmethods/lucaskanade.py

```diff
@@ -0,0 +1,94 @@
+    least squares problem to identify a change in parameters. We update the parameters and repeat.
+
+    To increase robustness, we extend the traditional LK algorithm to use a set of reference images.
+    We minimize the squared error of the difference of the transformed image and a learned weighting of the references.
+    """
+
+    def __init__(self, transformationType='Translation', border=0, tol=1e-5, maxIter=10, robust=False):
+        """
+        Parameters
+        ----------
+        transformationType : one of 'Translation', 'Euclidean', optional, default = 'Translation'
+            type of transformation to use
+        border : int or tuple, optional, default = 0
+            Border to be zeroed out after transformations. For most datasets, it is
+            critical that this value be larger than the maximum translation to get
+            good results and avoid boundary artifacts.
+        maxIter : int, optional, default = 10
+            maximum number of iterations
+        tol : float, optional, default = 1e-5
+            stopping criterion on the L2 norm of the change in parameters
+        robust : bool, optional, default = False
+            solve a least absolute deviation problem instead of least squares
+        """
+        self.transformationType = transformationType
+        self.border = border
+        self.maxIter = maxIter
+        self.tol = tol
+        self.robust = robust
+
+    def prepare(self, images, startidx=None, stopidx=None):
+        """
+        Prepare Lucas-Kanade registration by computing or specifying a reference image.
+
+        Parameters
+        ----------
+        images : ndarray or Images object
+            Images to compute reference from, or a single image to set as reference
+
+        See computeReferenceMean.
+        """
+        if isinstance(images, Images):
+            self.reference = computeReferenceMean(images, startidx, stopidx)
+        elif isinstance(images, ndarray):
+            self.reference = images
+        else:
+            raise Exception('Must provide either an Images object or a reference')
+        # Convert references to matrix to speed up solving linearized system
+        self.referenceMat = volumesToMatrix(self.reference)
+        return self
+
+    def isPrepared(self, images):
+        """
+        Check if Lucas-Kanade is prepared by checking the dimensions of the reference.
+
+        See checkReference.
+        """
+
+        if not hasattr(self, 'reference'):
+            raise Exception('Reference not defined')
+        else:
+            checkReference(self.reference, images)
+
+    def getTransform(self, vol):
+        from numpy.linalg import norm
+        # Create initial transformation
+        tfm = TRANSFORMATION_TYPES[self.transformationType](shift=zeros(vol.ndim))
+        # Grid of coordinates needed to transform images using map_coordinates
+        grid = GridTransformer(vol.shape)
```

Comment:
Can we define this in the `prepare` method? And have the `isPrepared` method check that it's been properly defined? `prepare` has access to the dimensions (via either the passed `Images` object, or the passed array). That way it's done once. As it is now, the task for every single image has to separately instantiate this object.